### PR TITLE
feat: implement RustGenerator proof-of-concept

### LINKED ISSUE #509: Implement Rust generator and renderers
Implement Rust generator and renderers like in TS or Java.

Generator should have possibility to define presets for struct, enum type.

Additional: create presets for serialisation/converters from json

### PR DESCRIPTION
**Pull request by leigh-johnson**
**Description**

Implements `RustGenerator` proof of concept with support for basic primitives, enum, structs, tuple, and union types. 

### Usage example

Check out `examples/rust-generate-crate` for a demo. You'll need to [install Rust](https://rustup.rs/) to compile the generated package. 

### Feature table

| **Feature**                  | **Status**    | **Info**                                                                              |
|------------------------------|---------------|---------------------------------------------------------------------------------------|
| Union (polymorphic type)     | ✅ done        |                                                                                       |
| Enum (group of constants)    | ✅ done        |                                                                                       |
| Array (unordered collection) | ✅ done        |                                                                                       |
| Tuple (ordered collection)   |  ✅ done       |                                                                                       |

### Next steps

* Implement To/From String methods (serde_json)
* Implement To/From u8 bytes methods (serde_bytes)
* Publish `asyncapi-nats-rs` repo for a full example of models in the wild

**Related issue(s)**
Fixes #509

### CODE REVIEW

File: src/generators/rust/renderers/TupleRenderer.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { RustRenderer } from '../RustRenderer';
+import { ConstrainedTupleModel, ConstrainedTupleValueModel } from '../../../models';
+import { TuplePresetType } from '../RustPreset';
+import { RustOptions } from '../RustGenerator';
+/**
+ * Renderer for Rust's `enum` type
```
Comment (jonaslagoni) [inline src/generators/rust/renderers/TupleRenderer.ts:6]:
```suggestion
 * Renderer for Rust's `Tuple` type
```

File: src/generators/rust/renderers/UnionRenderer.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { RustRenderer } from '../RustRenderer';
+import { ConstrainedUnionModel, ConstrainedMetaModel } from '../../../models';
+import { UnionPresetType } from '../RustPreset';
+import { RustOptions } from '../RustGenerator';
+
+/**
+ * Renderer for Rust's `enum` type
```
Comment (jonaslagoni) [inline src/generators/rust/renderers/UnionRenderer.ts:7]:
```suggestion
 * Renderer for Rust's `Union` type
```

File: examples/rust-generate-crate/index.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    ]
+  });
+  // Generated files will be written to output/ directory
+  const outDir = path.join(__dirname, 'output');
+  // Parse the schema document
+  const inputModel = await generator.process(doc);
+
+  // Run the file generator with options
+  const models = await generator.generateToFiles(inputModel, outDir, {
```
Comment (jonaslagoni) [inline examples/rust-generate-crate/index.ts:45]: You can simplify this if you wants:

```suggestion
  // Run the file generator with options
  const models = await generator.generateToFiles(doc, outDir, {
```

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Wow, that was damn quick and you even targeted `next` 🤩💪 Had a few comments and suggestions but overall looks great! 

Some extra changes which are needed:
- Please add rust to the list of generators: https://github.com/asyncapi/modelina/blob/next/README.md
- https://github.com/asyncapi/modelina/tree/next/docs#languages need to have rust added (I see I need to add Dart here as well)
- https://github.com/asyncapi/modelina/blob/next/docs/usage.md#generate-dart-models needs a basic rust usage example or just link to existing one.

If you want to continue to be engaged and take ownership of the rust generator feel free to add yourself as [code owner](https://github.com/asyncapi/modelina/blob/next/CODEOWNERS) in this PR 👍

**Comment by leigh-johnson** [comment]
Thank you for the thorough & thoughtful review! I blocked some time to check in revisions tomorrow (Friday) AM PDT.

:gem:  I should mention that I originally worked from `master` - so I really appreciate how much effort went into improving the generator development experience for `next`. The `MetaModel` refactor was excellent and made it a breeze to jump right in. Nice job y'all! 

> Wow, that was damn quick and you even targeted `next` star_struckmuscle Had a few comments and suggestions but overall looks great!
> 
> You can ignore all the CI checks, I am trying to solve them in another PR, so we merge regardless of failures.
> 
> Some extra changes which are needed:
> 
> * Please add rust to the list of generators: https://github.com/asyncapi/modelina/blob/next/README.md
> * https://github.com/asyncapi/modelina/tree/next/docs#languages need to have rust added (I see I need to add Dart here as well)
> * https://github.com/asyncapi/modelina/blob/next/docs/usage.md#generate-dart-models needs a basic rust usage example or just link to existing one.
> 
> If you want to continue to be engaged and take ownership of the rust generator feel free to add yourself as [code owner](https://github.com/asyncapi/modelina/blob/next/CODEOWNERS) in this PR +1

File: examples/rust-generate-crate/index.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      authors: ['AsyncAPI Rust Champions'],
+      homepage: 'https://www.asyncapi.com/tools/modelina',
+      repository: 'https://github.com/asyncapi/modelina',
+      license: 'Apache-2.0',
+      description: 'Rust models generated by AsyncAPI Modelina',
+      // support 2018 editions and up
+      edition: '2018',
+      // enable serde_json
+      packageFeatures: [RustPackageFeatures.json] as RustPackageFeatures[]
+    }
```
[the remark this answers is not in the document]
Comment (leigh-johnson) [reply examples/rust-generate-crate/index.ts:61]: I can cut the user-facing API to avoid scope creep.

We do need a `Cargo.toml` manifest and `lib.rs` module for black-box testing. Otherwise, compilation commands like `cargo build` and sanity checks like `cargo clippy` will fail. Would it make more sense to include this functionality in a test helper instead?

File: src/generators/rust/RustGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  constrainToMetaModel(model: MetaModel): ConstrainedMetaModel {
+    // How to constrain Union member names?
+    if (model instanceof UnionModel) {
+      model.union = model.union.map((v, idx) => {
+        v.name += idx;
+        return v;
+      });
+    }
```
[the remark this answers is not in the document]
Comment (leigh-johnson) [reply src/generators/rust/RustGenerator.ts:103]: Yeah, I have a feeling users will want to customize the way Union member names are generated. I couldn't do this in `ModelNameConstrainer.ts` because `ModelNameContext` only passes `modelName` to its fn partial context, but I need the full model to apply this naming constraint. 

In Rust, a Union type is implemented from an `enum` of all possible compositions - so the members need to be named. Here's a concrete example:

```typescript
// The `members` property is a union of one of: string, number, boolean
const doc = {
  $id: '_address',
  type: 'object',
  properties: {
    members: { oneOf: [{ type: 'string' }, { type: 'number' }, { type: 'boolean' }], },
}

// Typescript's expressive semantics allow us to define a union type in-place in the type declaration of Address
// members is composed from: string | number | boolean
class Address {
  private _members?: string | number | boolean;
  constructor(input: {
    members?: string | number | boolean,
  }) {
    this._marriage = input.marriage;
    this._members = input.members;
  }
  get members(): string | number | boolean | undefined { return this._members; }
  set members(members: string | number | boolean | undefined) { this._members = members; }
}
```

Here's the equivalent in Rust, which has much lower-level semantics for expressing that some memory value can be one of many types: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f81defc7eb7e45ecc85f17ebe3a91b60

The important bit is that Rust's union implementation needs a name for each possible type member.
```rust
// Members is a union type of a String, float, or bool value
#[derive(Clone, Debug, PartialEq)]
pub enum Members {
    Members0(String),
    Members1(f64),
    Members2(bool),
}
```

**Comment by leigh-johnson** [comment]
How do y'all prefer to solve merge conflicts? Let me know if you prefer I rebase and force-push, create a merge commit from next -> branch, or merge next -> branch with --no-ff

File: README.md [hunk 6]
```diff
@@ -123,6 +123,14 @@ To see the complete feature list for each language, please click the individual
     <td><a href="./docs/usage.md#generate-dart-models">Dart</a></td>
     <td>Class and enum generation: json_annotation</td>
   </tr>
+  <tr>
+    <td><a href="./docs/usage.md#generate-dart-models">Dart</a></td>
+    <td>Class and enum generation: json_annotation</td>
+  </tr>
```
Comment (jonaslagoni) [inline README.md:129]:
```suggestion
```

File: examples/rust-generate-crate/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply examples/rust-generate-crate/index.ts:59]: > We do need a Cargo.toml manifest and lib.rs module for black-box testing. Otherwise, compilation commands like cargo build and sanity checks like cargo clippy will fail. Would it make more sense to include this functionality in a test helper instead?

I think only you can answer that 😄 If you see it as a benefit that the user can generate the support files, then I am all good with keeping it 🙂

File: src/generators/rust/RustGenerator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/generators/rust/RustGenerator.ts:103]: That makes sense, thanks for the clarification 👍 Yea, have this on a backburner for other languages cause I dont know the best way to handle them as of yet (https://github.com/asyncapi/modelina/issues/393, https://github.com/asyncapi/modelina/issues/392, https://github.com/asyncapi/modelina/issues/391). So you are the first to have this need.

It's a bit of a conundrum, cause I am not sure `ModelNameConstrainer` make sense to place this logic in 🤔 Even if you had access to the union model in the context and ran this code, it would be overwritten once the constrainer walk over the union models next. So it will be somewhat of a mess 🙃 Need to think about that one...

I think we should keep it here for now and I will make another issue about this as a follow-up, how does that sound?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
> How do y'all prefer to solve merge conflicts? Let me know if you prefer I rebase and force-push, create a merge commit from next -> branch, or merge next -> branch with --no-ff

That is entirely up to you, it's your branch so you can do whatever you prefer 🙂 We are squashing commits when we merge, so nothing will be visible from Modelina's perspective.

Only had one small change otherwise it looks good 👍

File: CODEOWNERS [hunk 7]
```diff
@@ -41,4 +41,8 @@
 */generators/csharp @ron-debajyoti
 
 # Language Champions for Dart and it's presets
-*/generators/dart
\ No newline at end of file
+*/generators/dart
+
+# Language Champions for Rust and its presets
+*/generators/rust @leigh-johnson
```
Comment (leigh-johnson) [inline CODEOWNERS:47]: @jonaslagoni Just noticed this warning message on my end. Looks like I need to be added to the repo for this check to pass?

Screen Shot 2022-08-02 at 9 10 55 AM
 alt="Screen Shot 2022-08-02 at 9 10 48 AM" src="https://user-images.githubusercontent.com/2601819/182422378-c9071ea0-68cc-4d43-83db-a04bda2b31f1.png">

**Comment by leigh-johnson** [comment]
Just rebased & fixed the last lingering code smells. Thanks again for helping me polish up this PR and for all your work on AsyncAPI. 🙇‍♀️ 

Stay tuned for a Rust + NATS client generator next.✌️

File: CODEOWNERS (hunk 7 above)
Comment (jonaslagoni) [reply CODEOWNERS:47]: Yes, once merged I will add you as codeowner 🙂 So we can ignore this for now 😄

File: test/blackbox/blackbox.spec.ts [hunk 8]
```diff
@@ -33,6 +33,7 @@ const AsyncAPIV2_3Files = readFilesInFolder('AsyncAPI-2_3');
 const AsyncAPIV2_4Files = readFilesInFolder('AsyncAPI-2_4');
 
 const filesToTest = [
+<<<<<<< HEAD
```
Comment (jonaslagoni) [inline test/blackbox/blackbox.spec.ts:36]: You have some merge conflicts here that has not been solved 🙂

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
You removed a newline in the generator, so the test snapshots needs to be updated 🙂 And then the Blackbox tests have some remaining merge conflicts. 

Other then that it LGTM 🎉 

> Just rebased & fixed the last lingering code smells. Thanks again for helping me polish up this PR and for all your work on AsyncAPI. 
> Stay tuned for a Rust + NATS client generator next.

Thanks for taking the time to contribute! Cant wait 🎉

File: test/blackbox/blackbox.spec.ts (hunk 8 above)
Comment (leigh-johnson) [reply test/blackbox/blackbox.spec.ts:36]: Whoops, sorry about that! My editor usually raises a five-alarm fire if I miss merge conflicts, so I've gotten sloppy. :joy: Blocking out time to fix + retest.

**Comment by jonaslagoni** [comment]
You can ignore the code smells, those seem to be mine 🧐 

Regarding the bug, that seems like its relevant, or is it a false positive?

File: src/generators/rust/renderers/PackageRenderer.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      let mod = defaultModelNameConstraints()({ modelName });
+      mod = FormatHelpers.snakeCase(mod);
+      return `
+pub mod ${mod};
+pub use self::${mod}::*;`;
+    }).flat());
+
+    return `#[macro_use]
+extern crate serde;
+extern crate serde_json;
```
Comment (jonaslagoni) [inline src/generators/rust/renderers/PackageRenderer.ts:35]: Think my lack of Rust knowledge comes to play here and why I did not notice this before 😅 I think it's fine to keep it for now, I just want to make you aware of this 🙂 

I expect that in the future you will get request to remove this from the standard rendering, and instead move this to a `SerdePreset`. Don't have a full overview of whether this is even possible, depends on how tangled things are with each other I guess.

The reason is, generally, renderers just render the very basic model, with "no" dependencies and are the raw building blocks presets can extend i.e. with serde annotation.

This also enables us to all use the common renderers and then decide to add extra sprinkles on top, i.e. libraries to help serialize the models to JSON, XML or even bytes, or something completely different 🙂 

Similar raised issue for Java: https://github.com/asyncapi/modelina/issues/812

File: src/generators/rust/renderers/EnumRenderer.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    return renderer.defaultSelf();
+  },
+  item({ item }) {
+    if (item.value === 'HashMap<String, serde_json::Value>') {
+      return `${item.key}(${item.value})`;
+    }
+    return `${item.key}`;
+  },
+  itemMacro({ itemIndex, model }) {
+    const originalInput = model.originalInput.enum[itemIndex];
```
Comment (jonaslagoni) [inline src/generators/rust/renderers/EnumRenderer.ts:59]: Be careful about using `originalInput` in the renderers (even though you definitely can), as you assume the input is always JSON Schema, which definitely is the case for now, but in the future it wont be, as imagine the models are generated from Protobuf, or GraphQL Schemas 🙂

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
I think I forgot to run the CI before last review 🧐 My bad, as it seems to have caught some things. Added some suggestions how to fix the linter. 

Regarding the blackbox it's related to assuming `originalInput` has some value, which in some cases don't. Not sure on the specifics why it's the case here with JSON Schema inputs, have to debug the blackbox tests to figure that one out 🤔 

Unless you have clear idea about a fix, I am good with merging and then create issue to fix later as we are still in pre-release.

File: src/generators/rust/renderers/PackageRenderer.ts (hunk 9 above)
Comment (leigh-johnson) [reply src/generators/rust/renderers/PackageRenderer.ts:35]: Moving this to a preset in a later revision makes sense to me! For my own application, I have to support 2 ser/de strategies:

* https://crates.io/crates/serde_bytes for the raw output of an edge ML model
* https://docs.rs/serde_json/latest/serde_json/ for human-readable data

I'd like to put together 1 concrete example of a JSON-based NATs Rust application generated from an AsyncAPI schema, then circle back to the more abstract parts (like Serde Presets). 

There's quite a bit to chew on here when considering the transports! For example...

Rust's `nats.rs` crate brokers `bytes::Byte`, which is a low-level container for handling contiguous bits of memory (exactly what I expect from a transport layer). A `From<bytes::Byte`> implementation is required to serialize/deserialize this data structure into a model struct/enum/etc. 
https://github.com/nats-io/nats.rs/blob/main/async-nats/src/message.rs#L28

However, the `jetstream` enabled Nats.rs client does implement Serializable/Deserializble traits, plus the message body is a primitive u64 sequence (not `bytes::Byte` wrapper) https://github.com/nats-io/nats.rs/blob/main/async-nats/src/jetstream/stream.rs#L558

tldr: I'm still trying to figure out exactly where the boundaries for the Rust generator will be. I'm seeing a lot of possible permutations (below), but am just working on the `rust-nats-serde-json` story to simplify my path forward. 

* rust-nats-serde-<json|bytes|etc>
* rust-nats-jetstream-<json|bytes|etc>
* rust-nats-raw (no serde)

:open_mouth: :exploding_head:

Comment (jonaslagoni) [reply src/generators/rust/renderers/PackageRenderer.ts:35]: Definitely many options 😄 

Makes sense to me 👍

File: examples/rust-generate-crate/index.spec.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  test('and should log expected output to console', async () => {
+    await generate();
+    expect(spy.mock.calls.length).toEqual(6);
+    expect(spy.mock.calls[0]).toMatchSnapshot();
+    expect(spy.mock.calls[1]).toMatchSnapshot();
+    expect(spy.mock.calls[2]).toMatchSnapshot();
+    expect(spy.mock.calls[4]).toMatchSnapshot();
+    expect(spy.mock.calls[5]).toMatchSnapshot();
+    expect(spy.mock.calls[6]).toMatchSnapshot();
+
```
Comment (jonaslagoni) [inline examples/rust-generate-crate/index.spec.ts:16]:
```suggestion
```

File: src/generators/rust/RustFileGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+   * Generates all the models to an output directory each model with their own separate files. 
+   * 
+   * @param input
+   * @param outputDirectory where you want the models generated to
+   * @param options
+  */
+  public async generateToFiles(input: Record<string, unknown> | InputMetaModel, outputDirectory: string, options: RustRenderCompleteModelOptions): Promise<OutputModel[]> {
+    let generatedModels = await this.generateCompleteModels(input, options);
+    //Filter anything out that have not been successfully generated
+    generatedModels = generatedModels.filter((outputModel) => { return outputModel.modelName !== '' && outputModel.modelName !== undefined });
```
Comment (jonaslagoni) [inline src/generators/rust/RustFileGenerator.ts:18]:
```suggestion
    generatedModels = generatedModels.filter((outputModel) => { return outputModel.modelName !== '' && outputModel.modelName !== undefined; });
```

File: src/generators/rust/RustGenerator.ts [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    const inputModel = await this.processInput(input);
+    const manifestMetaModel = new ConstrainedSupportFileModel('Cargo.toml', inputModel, 'supportFile');
+    const libMetaModel = new ConstrainedSupportFileModel('src/lib.rs', inputModel, 'supportFile');
+    const manifestOutput = await this.renderManifest(manifestMetaModel, inputModel, options);
+    const libOutput = await this.renderLib(libMetaModel, inputModel, options);
+
+    return [
+      OutputModel.toOutputModel({
+        result: manifestOutput.result,
+        modelName: "CargoToml",
```
Comment (jonaslagoni) [inline src/generators/rust/RustGenerator.ts:192]:
```suggestion
      modelName: 'CargoToml',
```

File: src/generators/rust/RustGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        dependencies: manifestOutput.dependencies,
+        model: manifestMetaModel,
+        inputModel
+      }),
+      OutputModel.toOutputModel({
+        result: libOutput.result,
+        modelName: "lib",
```
Comment (jonaslagoni) [inline src/generators/rust/RustGenerator.ts:199]:
```suggestion
        modelName: 'lib',
```

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Should be the last suggestions for linter.

The blackbox tests fails, but lets fix it at another time, might need to spend some time debugging to get to the bottom of the issue.

**Review by jonaslagoni — APPROVED** [review]
🎉

**Comment by jonaslagoni** [comment]
@all-contributors please add @leigh-johnson for code, test, example, docs, maintenance

**Comment by jonaslagoni** [comment]
The team grows @magicmatatjahu @Samridhi-98 @ron-debajyoti 🎉 

@leigh-johnson you should have been invited to join as collaborator 😃 Welcome to the team! 

You are also illegible to join the AsyncAPI TSC (in case you didn't catch that 😄), all you have to do is add yourself to the list of TSC members here: https://github.com/asyncapi/community/blob/master/TSC_MEMBERS.json. An example of how to do it can be seen here: https://github.com/asyncapi/community/pull/430

Glad to have you onboard, and thanks for the patience with this PR 🍺
